src/waitlist.cpp

```diff
@@ -105,54 +106,42 @@ std::size_t WaitingList::getTime(std::size_t slot)
 	}
 }
 
-bool WaitingList::clientLogin(const Player* player)
+bool WaitingList::clientLogin(const Player* player, std::size_t& currentSlot)
```

Comment:
I personally really dislike this C-style function, I would make it return `std::optional<std::size_t>` or rework it so that `0` is an invalid return value that indicates failure. Not something to do right now though.

Reply:
This is a very educational experience for me (Im not a c/c++ dev, this PR is cherry picked), could you collaborate on the difference? It would be awesome if you could add a commit to this PR by pushing to the `fix-waiting-list` branch on `Znote/forgottenserver` so I can see how you would change this.

Reply:
What @ranisalt means by "C-style function" is that `WaitingList::clientLogin` returns a value, but also takes a value by reference as one of its parameters so that it can "return" more than one value. C doesn't have the concept of multiple return values (e.g., tuples), so devs would have to pass variables as arguments as sort of a hack.

With this C-style function, the caller **always** has to define a variable for `currentSlot` and pass it to `clientLogin`:
```cpp
std::size_t currentSlot;
bool canLogin = WaitingList::getInstance().clientLogin(player, currentSlot);
if (!canLogin) {
    // use the value set in currentSlot to get the retry time
    // currentSlot is only used if canLogin is false
}
```
A better way to do this would be to have `clientLogin` return a `std::optional<std::size_t>>` that you would then check if a value exists, or switch it from a `bool` return type to a `size_t` and there would be no need for the `currentSlot` parameter.

`std::optional`:
```cpp
std::optional<std::size_t> WaitingList::clientLogin(const Player* player)
{
    // do whatever...
    if (slot > 0) {
        return slot;
    }
    return std::nullopt;
}

{
    // there may be a prettier way to do this, but you get the idea
    auto ret = WaitingList::getInstance().clientLogin(player);
    if (ret) {
        auto currentSlot = ret.value();
        ....
    }
}
```
However, the code above checks if slot is 0 anyway, so you could just return slot and not use `std::optional`:
```cpp
size_t WaitingList::clientLogin(const Player* player)
{
    // do whatever...
    return slot;
}

{
    // there may be a prettier way to do this, but you get the idea
    auto currentSlot = WaitingList::getInstance().clientLogin(player);
    if (currentSlot != 0) {
        ....
    }
}
```

Reply:
Excellent explanation, @jo3bingham

Reply:
@jo3bingham Totally AWESOME! I think I'm starting to understand. Let me know what you think of my latest commit. I figured, why return a bool when you can just return currentSlot, and in scenarios where you skip current slot, return 0 as in (your not waiting for anyone, walk right in). That way, the function can be executed as a value to currentSlot in protocolgame.cpp, thus removing the need to pass anything through scetchy param references.